pkg/ddevapp/apptypes.go

```diff
@@ -0,0 +1,92 @@
+package ddevapp
+
+type settingsCreator func(*DdevApp) error
+type uploadDir func(*DdevApp) string
+
+// hookSuggestions should probably change its arg from string to app when
+// config refactor is done.
+type hookSuggestions func() []byte
+
+type siteSettingsPaths func(app *DdevApp)
+
+// AppTypeFuncs struct defines the functions that can be called (if populated)
+// for a given appType.
+type AppTypeFuncs struct {
+	settingsCreator
+	uploadDir
+	hookSuggestions
+	siteSettingsPaths
+}
+
+// AppTypeMatrix is a static map that defines the various functions to be called
+// for each apptype (CMS).
+// Example: AppTypeMatrix["drupal"]["7"] == { settingsCreator etc }
+var AppTypeMatrix = map[string]AppTypeFuncs{
+	"php": AppTypeFuncs{},
+	"drupal7": AppTypeFuncs{
+		CreateDrupalSettingsFile, GetDrupalUploadDir, GetDrupal7Hooks, SetDrupalSiteSettingsPaths,
+	},
+	"drupal8": AppTypeFuncs{
+		CreateDrupalSettingsFile, GetDrupalUploadDir, GetDrupal8Hooks, SetDrupalSiteSettingsPaths,
+	},
+	"wordpress": AppTypeFuncs{
+		CreateWordpressSettingsFile, nil, GetWordpressHooks, SetWordpressSiteSettingsPaths,
+	},
+	"backdrop": AppTypeFuncs{},
+	"typo3":    AppTypeFuncs{},
+}
+
+// GetValidAppTypes returns the valid apptype keys from the AppTypeMatrix
+func GetValidAppTypes() []string {
+	keys := make([]string, 0, len(AppTypeMatrix))
+	for k := range AppTypeMatrix {
+		keys = append(keys, k)
+	}
+	return keys
+}
+
+// IsValidAppType checks to see if the given apptype string is a valid configured
+// apptype.
+func IsValidAppType(apptype string) bool {
+	if _, ok := AppTypeMatrix[apptype]; ok {
+		return true
+	}
+	return false
+}
+
+// CreateSettingsFile creates the settings file (like settings.php) for the
+// provided app is the apptype has a settingsCreator function.
+func CreateSettingsFile(l *DdevApp) error {
+	if appFuncs, ok := AppTypeMatrix[l.GetType()]; ok && appFuncs.settingsCreator != nil {
+		err := appFuncs.settingsCreator(l)
+		return err
+	}
+	return nil
+}
+
+// UploadDirFunc returns the upload (public files) directory for the given app
+func UploadDirFunc(l *DdevApp) string {
```

Comment:
Is there ever a case where we need > 1 upload dir? For instance, if we eventually support Drupal multisite, do we need an upload dir for each site? Would it be worthwhile to make this a list of strings rather than a single value to make it easy to expand in the future?

Reply:
I dearly hope we'll never support any kind of multisite, don't think we will. 

However, we do have the problem of public and private. We currently have no support at all for Drupal's private files. 

I'd say this is probably out of scope for this PR, but it's a good place to think about it anyway.

Reply:
FWIW I just got a question about multisite yesterday. I'm in the same boat as @rfay though, I'd prefer not to support it.

Reply:
Multisite is so tremendously evil. I've never understood why in the world it cumbers the earth. I discovered back in 4.6 or something that you can't safely upgrade one of your multisites without breaking the others during the code upgrade, and it's never changed. Never have figured out what problem it solves (unless a MB of code is a storage problem for you).

Reply:
I don't particularly want to support multisite either, but it was an immediately obvious use for multiple upload dirs. Does changing this function's signature have a bigger impact? Because if it's a simple change, I think it's important to do this now so that we don't have a breaking change later down the road.

Reply:
My bet is the problem is not the func's signature, but rather our assumptions in code. You're right to question the assumption, but I don't think we should change it now because I'd rather aim at simple approaches that match our current thinking. For example, I'd expect we'll have a file_upload_dir in the config.yaml before very long at all, that this would return. If we start to make it plural ahead of the thinking, it implies things that aren't true. file_upload_dir could be a yaml array, but I'd rather see us grow into that in a later PR.

The way I see it right now, we're not committing to much with these signatures (although best to have them as good as they can be). It's not a lot of work to finesse for all the types they represent.